### ISSUE TITLE
QUESTION: APN device tokens are in a wrong (long) format

### ISSUE THREAD

**Issue by paulincai** [node-apn/node-apn#636]
IOS 12
phonegap-plugin-push@2.2.3
cordova-plugin-device@2.0.2
cordova@8.0.0
cordova-ios@4.5.5
apn@2.2.0
xCode10

I think I don't understand that whole token concept.

I do a **_PushNotification.init(options)_**.
On registration I receive an object { data } and **_data.registrationId_** is a very long token (+150 characters). I understand that I need a 64bit string in order to send notifications.

It looks that in Objective C the long string (probably a type of object) can be reduced to the correct string with this:  **_deviceToken.reduce('', {$0 + String(format: "%02X", $1)})_**

I also understand that when the xCode installs the app on a device, the xCode console will show the correct format of device token.

**In summary:**

**I get kind of this:** 'eSlF2WpGFnI:APA91bPCLnML1LBBG77mkBlHjrJLx-cHydupMYmrJ2HHHBzltFVeDo2ElZvjkSAQs38-y4zNGMfYaeKpNd3QMLUQfixYxhgtxIA0zGnOmsrEJucH4WhioWwJ49ato8t3EzA4YXQEfqK7'

**and I need kind of this:** 740f4707 bebcf74f 9b7c25d4 8e335894 5f6aa01d a5ddb387 462c7eaf 61bb78ad

Anyone can enlighten me please on how to get the correct format. 

When I send message to the long one I get "BadDeviceToken".
Thank you.

**Comment by paulincai** [comment]
Ok, confusion is clear now. My cordova push plugin is going the FCM route instead of APN. Those long registration ids are specific to FCM.

**Comment by johhansantana** [comment]
@paulincai how did you get the APN token? I'm using react-native-firebase but it also does use the FCM.

**Comment by paulincai** [comment]
@johhansantana I am not sure what exactly you ask. APN tokens are produced by Apple while FCM tokens are produced by Google. If you use FCM, you need FCM tokens.

**Comment by johhansantana** [comment]
🤔  well, this library is telling me that the FCM token that I'm getting from `react-native-firebase` is invalid and which is the same format that you mentioned above. So I'm assuming that I would just need to get the APN token instead of the FCM that the `react-native-firebase` provides. I was wondering how did you solve this.

**Comment by paulincai** [comment]
"this library is telling me that the FCM token that I'm getting from react-native-firebase is invalid " - somehow I feel you answer this yourself. If you use APN (node-apn) to send to IOS and to register IOS devices, you can only use FCM for Android. You should not use react-native-firebase to generate device token because they are incompatible with APN. Consider APN and FCM as gateways, if you generate a token with one, you need to send Push via the same gateway. For Android things are clear, you need to use FCM but for IOS you can use both FCM and APN. You just need to decide for one and if you opt for APN, you need to block react-native-firebase from interacting with IOS.
You either do the IOS setup (https://rnfirebase.io/#3-ios-setup) and go with Firebase or you don't do the setup and go via APN. What you decide to do client side, you need to support on your server. My suggestion, unless you have specific requirement that only APN can satisfy, better use FCM for bot IOS and Android and let go this library.

**Comment by paulincai** [comment]
@johhansantana all ok? On the FCM I could help you with some pieces of code for Node (server side)

**Comment by johhansantana** [comment]
> @johhansantana all ok? On the FCM I could help you with some pieces of code for Node (server side)

I was trying to make the apple wallet pass to update via push notification but after further reading I saw that the wallet app will send you to your API the device APN token and was able to use that. It seems to work correctly with this now. 

Thank you for explaining the difference.
